libvast/src/transform_steps/select.cpp

```diff
@@ -0,0 +1,75 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2021 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "vast/transform_steps/select.hpp"
+
+#include "vast/concept/parseable/to.hpp"
+#include "vast/concept/parseable/vast/expression.hpp"
+#include "vast/error.hpp"
+#include "vast/expression.hpp"
+#include "vast/logger.hpp"
+#include "vast/plugin.hpp"
+#include "vast/table_slice_builder_factory.hpp"
+
+#include <arrow/type.h>
+#include <caf/expected.hpp>
+
+namespace vast {
+
+select_step::select_step(std::string expr)
+  : expression_(caf::none) { // FIXME: Is caf::none OK?
+  auto e = to<vast::expression>(std::move(expr));
+  if (!e) {
+    expression_ = e;
+    return;
+  }
+  // FIXME: Is normalize useful here?
+  expression_ = normalize_and_validate(*e);
+}
+
+caf::expected<table_slice> select_step::operator()(table_slice&& slice) const {
+  if (!expression_)
+    return expression_.error(); // FIXME: Is this message OK?
+  auto tailored_expr = tailor(*expression_, slice.layout());
+  if (!tailored_expr)
+    return tailored_expr.error();
+  auto new_slice = filter(slice, *tailored_expr);
+  if (new_slice)
+    return *new_slice;
+  // FIXME: return an empty slice or caf::error if new_slice is empty?
```

Comment:
I think currently we don't allow empty table slices, so it will have to be a `caf::error`.